File: react-native/src/tracing/msrSpanBuilder.ts
import type { IIdProvider } from "../utils/idProvider";
import type { Logger } from "../utils/logger";
import type { TimeProvider } from "../utils/timeProvider";
import { MsrSpan } from "./msrSpan";
import type { Span } from "./span";
import type { SpanBuilder } from "./spanBuilder";
import type { ISpanProcessor } from "./spanProcessor";
import type { ITraceSampler } from "./traceSampler";

/**
 * Concrete implementation of the SpanBuilder protocol.
 * It holds the necessary dependencies and configuration to create a new MsrSpan.
 */
export class MsrSpanBuilder implements SpanBuilder {
    private readonly name: string;
    private readonly idProvider: IIdProvider;
    private readonly timeProvider: TimeProvider;
    private readonly spanProcessor: ISpanProcessor;
    private readonly traceSampler: ITraceSampler;
    private readonly logger: Logger;
    
    private parentSpan?: Span;

    constructor(
        name: string,
        idProvider: IIdProvider,
        timeProvider: TimeProvider,
        spanProcessor: ISpanProcessor,
        traceSampler: ITraceSampler,
        logger: Logger
    ) {
        this.name = name;
        this.idProvider = idProvider;
        this.timeProvider = timeProvider;
        this.spanProcessor = spanProcessor;
        this.traceSampler = traceSampler;
        this.logger = logger;
    }

    /**
     * Sets the parent span for the span being built.
     * @param span The span to set as parent
     * @returns The builder instance for method chaining
     */
    public setParent(span: Span): SpanBuilder {
        this.parentSpan = span;
        return this;
    }

    /**
     * Creates and starts a new span with the current time.
     * @returns A new Span instance
     */
    public startSpan(): Span;

    /**
     * Creates and starts a new span with the specified start time.
     * @param timestampMs The start time in milliseconds since epoch
     * @returns A new Span instance
     */
    public startSpan(timestampMs: number): Span;

    /**
     * Implementation for startSpan overloads.
     */
    public startSpan(timestampMs?: number): Span {
        return MsrSpan.startSpan({
            name: this.name,
            logger: this.logger,
            timeProvider: this.timeProvider,
            idProvider: this.idProvider,
            traceSampler: this.traceSampler,
            parentSpan: this.parentSpan,
            spanProcessor: this.spanProcessor,
            timestamp: timestampMs,
        });
    }
}

File: ios/Sources/MeasureSDK/Swift/Tracing/MsrSpanBuilder.swift
import Foundation

/// A builder class for creating and configuring new spans.
final class MsrSpanBuilder: SpanBuilder {
    private let name: String
    private let idProvider: IdProvider
    private let timeProvider: TimeProvider
    private let spanProcessor: SpanProcessor
    private let sessionManager: SessionManager
    private let signalSampler: SignalSampler
    private let logger: Logger
    private var parentSpan: Span?

    init(name: String,
         idProvider: IdProvider,
         timeProvider: TimeProvider,
         spanProcessor: SpanProcessor,
         sessionManager: SessionManager,
         signalSampler: SignalSampler,
         logger: Logger) {
        self.name = name
        self.idProvider = idProvider
        self.timeProvider = timeProvider
        self.spanProcessor = spanProcessor
        self.sessionManager = sessionManager
        self.signalSampler = signalSampler
        self.logger = logger
    }

    /// Sets the parent span for the span being built.
    /// - Parameter span: The span to set as parent
    /// - Returns: The builder instance for method chaining
    func setParent(_ span: Span) -> SpanBuilder {
        self.parentSpan = span
        return self
    }

    /// Creates and starts a new span with the current time.
    /// - Returns: A new Span instance
    func startSpan() -> Span {
        return MsrSpan.startSpan(name: name,
                                 logger: logger,
                                 timeProvider: timeProvider,
                                 sessionManager: sessionManager,
                                 idProvider: idProvider,
                                 signalSampler: signalSampler,
                                 parentSpan: parentSpan,
                                 spanProcessor: spanProcessor)
    }

    /// Creates and starts a new span with the specified start time.
    /// - Parameter timeMs: The start time in milliseconds since epoch
    /// - Returns: A new Span instance
    func startSpan(_ timestamp: Number) -> Span {
        return MsrSpan.startSpan(name: name,
                                 logger: logger,
                                 timeProvider: timeProvider,
                                 sessionManager: sessionManager,
                                 idProvider: idProvider,
                                 signalSampler: signalSampler,
                                 parentSpan: parentSpan,
                                 spanProcessor: spanProcessor,
                                 timestamp: timestamp)
    }
}
